File: KeybondKeyboard/WordPredictor.swift
import Foundation

/// Simple on-device, prefix-based word predictor. No network access, no
/// UITextChecker dependency — just a bundled frequency-ordered word list
/// per language plus a small in-session boost for recently typed words.
final class WordPredictor {
    private var wordsByLanguage: [KeyboardLanguage: [String]] = [:]
    private var sessionBoost: [String: Int] = [:]

    init() {
        wordsByLanguage[.english] = Self.loadWords(resource: "words_en")
        wordsByLanguage[.thai] = Self.loadWords(resource: "words_th")
    }

    private static func loadWords(resource: String) -> [String] {
        guard let url = Bundle.main.url(forResource: resource, withExtension: "txt"),
              let contents = try? String(contentsOf: url, encoding: .utf8) else {
            return []
        }
        var seen = Set<String>()
        var ordered: [String] = []
        for raw in contents.split(separator: "\n") {
            let word = String(raw).trimmingCharacters(in: .whitespaces)
            guard !word.isEmpty, !seen.contains(word) else { continue }
            seen.insert(word)
            ordered.append(word)
        }
        return ordered
    }

    /// Returns up to `limit` suggestions for the given prefix, ranked by
    /// session usage first, then dictionary (frequency) order.
    func suggestions(for prefix: String, language: KeyboardLanguage, limit: Int = 3) -> [String] {
        let words = wordsByLanguage[language] ?? []
        guard !prefix.isEmpty else {
            // Nothing typed yet: show the most common words, like iOS does
            // before you start a sentence.
            return Array(words.prefix(limit))
        }
        let lowerPrefix = prefix.lowercased()

        let matches = words.filter { $0.lowercased().hasPrefix(lowerPrefix) }
        let ranked = matches.sorted { a, b in
            let boostA = sessionBoost[a.lowercased()] ?? 0
            let boostB = sessionBoost[b.lowercased()] ?? 0
            if boostA != boostB { return boostA > boostB }
            return false // keep dictionary (frequency) order otherwise
        }
        return Array(ranked.prefix(limit))
    }

    /// Call when the user commits a word (types a space/punctuation after it)
    /// so it ranks higher next time.
    func recordUsage(of word: String) {
        let key = word.lowercased()
        guard !key.isEmpty else { return }
        sessionBoost[key, default: 0] += 1
    }
}

File: KeybondAndroid/app/src/main/java/com/keybond/android/ime/WordPredictor.kt
package com.keybond.android.ime

import android.content.Context
import com.keybond.android.R
import java.io.BufferedReader
import java.io.InputStreamReader

/**
 * Simple on-device, prefix-based word predictor. No network access — just a
 * bundled frequency-ordered word list per language plus a small in-session
 * boost for recently used words.
 */
class WordPredictor(context: Context) {

    private val wordsByLanguage: Map<KeyboardLanguage, List<String>> = mapOf(
        KeyboardLanguage.ENGLISH to loadWords(context, R.raw.words_en),
        KeyboardLanguage.THAI to loadWords(context, R.raw.words_th)
    )

    private val sessionBoost = mutableMapOf<String, Int>()

    private fun loadWords(context: Context, resId: Int): List<String> {
        val seen = LinkedHashSet<String>()
        context.resources.openRawResource(resId).use { stream ->
            BufferedReader(InputStreamReader(stream, Charsets.UTF_8)).useLines { lines ->
                lines.forEach { raw ->
                    val word = raw.trim()
                    if (word.isNotEmpty()) seen.add(word)
                }
            }
        }
        return seen.toList()
    }

    /**
     * Returns up to [limit] suggestions for the given prefix, ranked by
     * session usage first, then dictionary (frequency) order.
     */
    fun suggestions(prefix: String, language: KeyboardLanguage, limit: Int = 3): List<String> {
        val words = wordsByLanguage[language].orEmpty()
        if (prefix.isEmpty()) {
            // Nothing typed yet: show the most common words, like a native
            // keyboard does before you start a sentence.
            return words.take(limit)
        }
        val lowerPrefix = prefix.lowercase()
        val matches = words.filter { it.lowercase().startsWith(lowerPrefix) }
        val ranked = matches.sortedByDescending { sessionBoost[it.lowercase()] ?: 0 }
        return ranked.take(limit)
    }

    /** Call when the user commits a word (types a space/punctuation after it). */
    fun recordUsage(word: String) {
        val key = word.lowercase()
        if (key.isEmpty()) return
        sessionBoost[key] = (sessionBoost[key] ?: 0) + 1
    }
}
